seed-job/index_reader.py

```diff
@@ -0,0 +1,369 @@
+        """
+        command = "oc get bc -o name -n {}".format(self.namespace)
+        bcs = run_cmd(command)
+        if not bcs.strip():
+            return []
+        else:
+            return bcs.strip().split("\n")
+
+    def oc_apply_seedjob_template_command(
+            self, template_location="seed-job/template.yaml"):
+        """
+        Returns the oc process and oc apply commands with parameters
+        to process seed-job/template.yaml
+        """
+        oc_apply = "oc apply -n {} -f -".format(self.namespace)
+        oc_process = "oc process -f {0} {1}".format(
+            template_location,
+            self.seed_template_params
+        )
+        return oc_process + "|" + oc_apply
+
+    def apply_buildconfigs(self, project):
+        """
+        Given a project object representing a project in container index,
+        process the seed-job template for same and oc apply changes
+        """
+        # get the oc process and oc apply command
+        command = self.oc_apply_seedjob_template_command()
+        # format the command with project params
+        command = command.format(
+            git_url=project.git_url,
+            git_path=project.git_path,
+            git_branch=project.git_branch,
+            target_file=project.target_file,
+            build_context=project.build_context,
+            desired_tag=project.desired_tag,
+            depends_on=project.depends_on,
+            notify_email=project.notify_email,
+            pipeline_name=project.pipeline_name,
+            app_id=project.app_id,
+            job_id=project.job_id,
+            pre_build_context=project.pre_build_context,
+            pre_build_script=project.pre_build_script,
+            registry_url=self.registry_url,
+            from_address=self.from_address,
+            smtp_server=self.smtp_server
+        )
+        # process and apply buildconfig
+        output = run_cmd(command, shell=True)
+        print (output)
+
+        # if a buildConfig has config update, oc apply returns
+        # "buildconfig.build.openshift.io "$PIPELINE_NAME" configured"
+        # possible values are ["unchanged", "created", "configured"]
+        # we are looking for "configured" string for updated pipelines
+        if "configured" in output:
+            print ("{} is updated, starting build..".format(
+                project.pipeline_name))
+            self.start_build(project.pipeline_name)
+
+    def start_build(self, pipeline_name):
+        """
+        Given a pipeline name, start the build for same
+        """
+        command = "oc start-build {} -n {}".format(
+            pipeline_name, self.namespace)
+        print (run_cmd(command))
+
+    def delete_buildconfigs(self, bcs):
+        """
+        Deletes the given list of bcs
+        """
+        command = "oc delete -n {} bc {}"
+        for bc in bcs:
+            print ("Deleting buildConfig {}".format(bc))
+            run_cmd(command.format(self.namespace, bc))
```

Comment:
I have experienced that few times, doing `oc delete bc/blah` doesn't work on first attempt. It doesn't throw error/warning/info. It just doesn't delete the buildconfig. I've never tried to dig for the reason. Maybe I should. Maybe it's a non-blocking call in that, given buildconfig gets deleted in background without blocking the command execution. But that's a guess. Should we take into account this scenario and add some check to ensure that the buildconfig we're trying to delete here has indeed been deleted? 

WDYT @bamachrn @navidshaikh

Reply:
Say, as per your experiences with oc delete, if we invoked `oc delete` and it logged the request and working on it in background (which eventuall would be completed I hope?) , what index reader should do, wait until it gets deleted?

Also, for cases as you mentioned, does oc print something on stdout?

Reply:
> (which eventuall would be completed I hope?)

TBH, I haven't been patient enough with it and have triggered `oc delete` again without letting much time pass by.

>  what index reader should do, wait until it gets deleted?

AFAIK, the call to `oc delete bc/blah` is a non-blocking one as I mentioned earlier. So we can't really make the index reader wait till it gets deleted. We could possibly add a check that ensures that given bc is deleted. If not, we trigger that `oc delete` command again.

> Also, for cases as you mentioned, does oc print something on stdout?

Nope, nothing.

Reply:
If its a non-blocking, I don't think we should re-run the command. I never encountered this, may be if we can re-produce this, we'll have better understanding of the scenario to write provision for it.